### ISSUE TITLE
After upgrading Scheduler to 3.3.0 Google Nest does not change to Heat according to the scheduler any more (not responding to the scheduler).

### ISSUE THREAD

**Issue by 730522js** [nielsfaber/scheduler-component#328]
### Checklist
- [x] I checked for similar existing issues (both open and closed) before posting.
- [x] I am running the latest version.

### Expected behavior
Google Nest to change mode from Eco to Heat according to the scheduler

### Actual behavior
Google Nest does not change the mode to Heat according to the scheduler, it remains on Eco mode.

I have scheduler active for other climate entities (Daiking heat pumps) and those change their modes according to the Scheduler as expected.
I've tested manually and HA is able to control the Google Nest thermostat (no connectivity issue).
I haven't changed anything before or after upgrading to Scheduler 3.3.0.
I've tried: 
- changing the scheduler (Time settings) from Daily occurrence to Choose (and manually selected each day of the week) => no change in behaviour. 
- changing the set temperature during the time it should be heating => no change in behaviour
- changing the Action from **Heat** to **Set Preset: none** => This actually works, so Scheduler can change the mode of Google Nest from Eco to None (this causes the Google Nest to resume last set temperature before it was switched to Eco) so I use this for the moment as a workaround to still have my scheduled temperatures across the hours and days.
- bottom line is: Scheduler can no longer command Google Nest after upgrading to 3.3.0 to change mode to Heat

### Steps to Reproduce
[Explain how to see the behaviour]

**Comment by 730522js** [comment]
I've deleted the Schedule for Google Nest and created it from scratch with the same time constrains and specifications and so far seems to work.

**Comment by etamtlosz** [comment]
Similar issue here with my Daikin. My schedule is to have the house on vent mode until midnight, then off. In the morning it should change from off to heating and 22C.
At 6:30 the heat turns on, but within a second it jumps to fan only mode, or to whatever mode it was in previously.

I don't think is just Nest related.

<img width="235" alt="image" src="https://github.com/nielsfaber/scheduler-component/assets/7436784/872e49c4-4dd1-4cb8-a4fa-cffffe8293b8">

**Comment by nielsfaber** [comment]
What scheduler does is:
1. Set climate entity operation mode to 'heat' (if it is not in mode 'heat') (service `climate.set_hvac_mode`)
2. Wait until entity has updated its state to 'heat' (maximum 50 seconds).
3. Set temperature setting to the chosen value (service `climate.set_temperature`)

I wouldn't know what is wrong about this that could explain the issues. And I don't see any relation with preset mode or fan mode (this behaviour seems specific to your devices).

I created a script that mimics the behaviour of scheduler (above steps).
I would like you ask you to test this in combination with your thermostats and check if it works.
The script will log some information in the home assistant log.
The code below can be added to the scripts.yaml file (or paste via the UI editor in the yaml mode).

```yaml
thermostat_test:
  variables:
    entity: climate.my_thermostat # insert entity ID here
    temperature: 20 # choose a setpoint different than current value
    mode: heat
  sequence:
    - service: system_log.write
      data:
        message: "-- Start of script --"
    - choose:
        - conditions:
            - condition: template
              value_template: "{{ is_state(entity, mode) }}"
          sequence:
            - service: system_log.write
              data:
                message: Mode is already {{ mode }}, skipping step
      default:
        - service: system_log.write
          data:
            message: Setting mode to {{ mode }}
        - service: climate.set_hvac_mode
          data_template:
            entity_id: "{{ entity }}"
            hvac_mode: "{{ mode }}"
    - service: system_log.write
      data:
        message: Waiting until {{ entity }} has state {{ mode }}
    - wait_template: "{{ is_state(entity, mode) }}"
      timeout: "00:00:50"
    - service: system_log.write
      data:
        message: Wait for state update completed.
    - choose:
        - conditions:
            - condition: template
              value_template: "{{ state_attr(entity, 'temperature') == temperature }}"
          sequence:
            - service: system_log.write
              data:
                message: Temperature is already {{ temperature }}, skipping step
      default:
        - service: system_log.write
          data:
            message: Setting temperature to {{ temperature }}
        - service: climate.set_temperature
          data_template:
            entity_id: "{{ entity }}"
            temperature: "{{ temperature }}"
    - service: system_log.write
      data:
        message: "-- End of script --"
  alias: test
  mode: single
```
Please post your findings.

**Comment by etamtlosz** [comment]
This has been working flawlessly during the last few years, that's why I assumed it is integration related.
Weird things started happening after the latest update.

Here's the output of the script
![2024-01-31-11-26-030PC2EChLlh](https://github.com/nielsfaber/scheduler-component/assets/7436784/2e5d1919-2b49-4fd3-a19a-df70cef5b05f)

**Comment by nielsfaber** [comment]
@etamtlosz The script shows the same behaviour as scheduler so that's good, we can reproduce the issue.
From the screen capture it appears that the first command (setting the mode to heat) is processed OK.
Somehow your climate device seems to jump back to fan mode when changing the setpoint.
Your test proves that the behaviour is caused by your device itself.
If you would use the dashboard card to perform the same steps, you will likely see the same behaviour.

I don't own the same device so from here I don't know how to make it work. 
I suggest you to make changes to the script until the response is as it should be.
Things to try:
1. Commenting out the setting of the temperature, to see if only setting operation mode works
2. Adding a small fixed time delay between the steps
3. Switching the order, so first setting the setpoint, then the mode
4. Setting the mode+setpoint simultaneously, so removing the time delay

**Comment by etamtlosz** [comment]
@730522js can you confirm your situation is similar? I don't want to hijack this issue. If yours something else, please report and I'll open another issue.

**Comment by 730522js** [comment]
> @730522js can you confirm your situation is similar? I don't want to hijack this issue. If yours something else, please report and I'll open another issue.

Yes, I can confirm my situation is similar.

**Comment by etamtlosz** [comment]
1. Works to set just the mode, it won't revert to the  previous mode.
2. Adding a 1 second delay seems to work
3. Messes up things, sometimes the temperature set is working but the mode reverts
4. Didn't manage to test this option

**Comment by peteh** [comment]
I have the same issue with my heaters. 
I am not explicitly setting a temperature though. 

My heaters take about 3-5 seconds the change the mode after selecting a different one.

**Comment by peteh** [comment]
Afaik still not resolved

**Comment by peteh** [comment]
Looks like when switching to auto somehow it requires target temperature also for AUTO

```
2024-03-15 15:31:53.719 ERROR (MainThread) [homeassistant.util.logging] Exception in async_timer_updated when dispatching 'scheduler_timer_updated': ('363511',)
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/site-packages/voluptuous/schema_builder.py", line 272, in __call__
    return self._compiled([], data)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/voluptuous/schema_builder.py", line 818, in validate_callable
    return schema(data)
           ^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/config_validation.py", line 171, in validate
    raise vol.Invalid(f"must contain at least one of {expected}.")
voluptuous.error.Invalid: must contain at least one of temperature, target_temp_high, target_temp_low.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/config/custom_components/scheduler/switch.py", line 251, in async_timer_updated
    await self._action_handler.async_queue_actions(
  File "/config/custom_components/scheduler/actions.py", line 257, in async_queue_actions
    await queue.async_start(skip_initial_execution)
  File "/config/custom_components/scheduler/actions.py", line 389, in async_start
    await self.async_process_queue()
  File "/config/custom_components/scheduler/actions.py", line 589, in async_process_queue
    await async_call_from_config(
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 291, in async_call_from_config
    await hass.services.async_call(**params, blocking=blocking, context=context)
  File "/usr/src/homeassistant/homeassistant/core.py", line 2284, in async_call
    processed_data: dict[str, Any] = handler.schema(service_data)
                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/voluptuous/validators.py", line 232, in __call__
    return self._exec((Schema(val) for val in self.validators), v)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/voluptuous/validators.py", line 355, in _exec
    raise e if self.msg is None else AllInvalid(self.msg, path=path)
  File "/usr/local/lib/python3.12/site-packages/voluptuous/validators.py", line 351, in _exec
    v = func(v)
        ^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/voluptuous/schema_builder.py", line 276, in __call__
    raise er.MultipleInvalid([e])
voluptuous.error.MultipleInvalid: must contain at least one of temperature, target_temp_high, target_temp_low.

```

**Comment by nielsfaber** [comment]
Scheduler only executes a service in HA with given set of parameters.
So if an action does not work with your device, the best way to identify the issue is by knowing what service configuration is needed for your thermostat.
The actions shown in the card are calculated based on the capabilities advertised by the climate entity, perhaps something is wrong here.
I suggest to use the developer tools -> services page to manually try putting your thermostat in the correct setting.
You can also check what is done by scheduler by enable the debug logging for the integration. After doing so, you should see extra lines appearing in the HA log at the time the schedule triggers.
Please share your findings, it helps me identifying the issue and fix this.

**Comment by peteh** [comment]
Alright, playing around a bit more, I'm using time slots. 
SET AUTO, no temp, current state is off: 
```
2024-03-17 22:34:47.477 DEBUG (MainThread) [custom_components.scheduler.switch] Schedule 363511 is starting in a timeslot, proceed with actions
2024-03-17 22:34:47.477 DEBUG (MainThread) [custom_components.scheduler.actions] [363511]: Executing service climate.set_temperature on entity climate.office_heater
2024-03-17 22:34:47.598 ERROR (MainThread) [homeassistant.util.logging] Exception in async_timer_updated when dispatching 'scheduler_timer_updated': ('363511',)
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/site-packages/voluptuous/schema_builder.py", line 272, in __call__
    return self._compiled([], data)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/voluptuous/schema_builder.py", line 818, in validate_callable
    return schema(data)
           ^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/config_validation.py", line 171, in validate
    raise vol.Invalid(f"must contain at least one of {expected}.")
voluptuous.error.Invalid: must contain at least one of temperature, target_temp_high, target_temp_low.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/config/custom_components/scheduler/switch.py", line 251, in async_timer_updated
    await self._action_handler.async_queue_actions(
  File "/config/custom_components/scheduler/actions.py", line 259, in async_queue_actions
    await queue.async_start(skip_initial_execution)
  File "/config/custom_components/scheduler/actions.py", line 406, in async_start
    await self.async_process_queue()
  File "/config/custom_components/scheduler/actions.py", line 606, in async_process_queue
    await async_call_from_config(
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 291, in async_call_from_config
    await hass.services.async_call(**params, blocking=blocking, context=context)
  File "/usr/src/homeassistant/homeassistant/core.py", line 2284, in async_call
    processed_data: dict[str, Any] = handler.schema(service_data)
                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/voluptuous/validators.py", line 232, in __call__
    return self._exec((Schema(val) for val in self.validators), v)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/voluptuous/validators.py", line 355, in _exec
    raise e if self.msg is None else AllInvalid(self.msg, path=path)
  File "/usr/local/lib/python3.12/site-packages/voluptuous/validators.py", line 351, in _exec
    v = func(v)
        ^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/voluptuous/schema_builder.py", line 276, in __call__
    raise er.MultipleInvalid([e])
voluptuous.error.MultipleInvalid: must contain at least one of temperature, target_temp_high, target_temp_low.
```

SET AUTO, temperature: 21deg

- just works

## From Dev console: 
set hvac mode to auto on office heater device -> works
```
service: climate.set_hvac_mode
target:
  device_id: 7706ce9490bae4a3d1e15b7c607273e6
data:
  hvac_mode: auto
```

set auto mode on entity id --> does not work
```
service: climate.set_hvac_mode
target:
  entity_id: climate.office_heater
data:
  hvac_mode: auto

```

**Comment by nielsfaber** [comment]
@peteh Thanks for your help investigating this!
I can confirm that scheduler was not properly handling the case of setting the mode to auto without setting the temperature (i.e. leaving the checkbox empty).
This would result in calling the service `climate.set_temperature` with no temperature setting but only `hvac_mode` parameter, which is not allowed in HA.
I made a fix for this, in this specific case the service `climate.set_hvac_mode` will now be executed. This fix is released under v3.3.2.

I don't think this solves the bug that was originally reported here, @730522js / @etamtlosz please state if you're still experiencing issues and what solution/workaround would be helpful. It is not clear to me how to facilitate the support for the Google Nest specifically.

**Comment by 730522js** [comment]
Hi all.

For me I can confirm I experience the behaviour I've initially opened the topic for.
I'm still going on the workaround I found (described further below).

So main issue (which was not there before 3.3.0):
- If I have the Google Nest thermostat set to Eco via ACTION > SET PRESET > ECO say from 8AM to 4PM and then the next action would be (as from 4PM) ACTION > HEAT > 22 degrees, then nothing happens, the Google Nest remains on Eco.
- If I manually change the Google Nest to Heat on 22 degrees, the use the Scheduler to modify the temperature...say from 10PM ACTION > HEAT > 21 degrees, then it works. 

Workaround:
- When I want to switch from Eco to Heating, I set an intermediate action in between Eco and Heat to Set Preset NONE.
Example below: 
- 8AM - 4PM > ACTION > SET PRESET > ECO
- 4PM - 4:05PM > ACTION > SET PRESET > NONE
- 4:05PM - 10PM > ACTION > HEAT > 22 degrees

So I can conclude:
- ACTION: Set Preset to ECO then next schedule is changing to HEAT = does not work (Google Nest remains on Eco)
- ACTION: Heat (any temperature) then next schedule Set Preset Eco or to HEAT (any temperature) = works
- ACTION: Set Preset NONE, then next schedule Action HEAT (any temperature) = works
<img width="1735" alt="2" src="https://github.com/nielsfaber/scheduler-component/assets/116653063/31cc773d-4695-4bd0-840f-f65a4cb68931">
<img width="1735" alt="1" src="https://github.com/nielsfaber/scheduler-component/assets/116653063/d6c8d04a-2c07-47da-833d-8fe96c794246">
